Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/approximated_centroidal_Voronoi_diagram_remeshing.h

```diff
@@ -0,0 +1,1385 @@
+  // TODO: (possible optimization to save memory) if qem is not used use std::conditionnal to store and compute less things in QEMClusterData
+  // TODO: (possible optimization to save memort) use a std::array to store quadrics and assemble the Eigen matric only at the end
+
+  std::size_t nb_vertices = vertices(tmesh).size();
+
+  // For remeshing, we might need to subdivide the mesh before clustering
+  // This should always hold: nb_clusters <= nb_vertices * vertex_count_ratio
+  // So do the following until the condition is met
+  if (gradation_factor == 0 && nb_clusters > nb_vertices * vertex_count_ratio)
+  {
+    std::vector<FT> lengths;
+    lengths.reserve(num_edges(tmesh));
+    FT cum = 0;
+    int nbe = 0;
+    for (edge_descriptor e : edges(tmesh))
+    {
+      lengths.push_back(edge_length(halfedge(e, tmesh), tmesh, parameters::vertex_point_map(vpm).geom_traits(gt)));
+      cum += lengths.back();
+      ++nbe;
+    }
+    FT threshold = 3. * cum / nbe;
+
+    std::vector<std::pair<edge_descriptor, double>> to_split;
+    int ei = -1;
+    for (edge_descriptor e : edges(tmesh))
+    {
+      FT l = lengths[++ei];
+      if (l >= threshold)
+      {
+        double nb_subsegments = std::ceil(l / threshold);
+        to_split.emplace_back(e, nb_subsegments);
+      }
+    }
+
+    while (!to_split.empty())
+    {
+      std::vector<std::pair<edge_descriptor, double>> to_split_new;
+      for (auto [e, nb_subsegments] : to_split)
+      {
+        halfedge_descriptor h = halfedge(e, tmesh);
+        Point_3 s = get(vpm, source(h,tmesh));
+        Point_3 t = get(vpm, target(h,tmesh));
+
+        for (double k=1; k<nb_subsegments; ++k)
+        {
+          // the new halfedge hnew pointing to the inserted vertex.
+          // The new halfedge is followed by the old halfedge, i.e., next(hnew,g) == h.
+          halfedge_descriptor hnew = Euler::split_edge(h, tmesh);
+          put(vpm, target(hnew, tmesh), barycenter(t, k/nb_subsegments, s));
+          for (halfedge_descriptor hhh : {hnew, opposite(h, tmesh)})
+          {
+            if (!is_border(hhh, tmesh))
+            {
+              halfedge_descriptor hf = Euler::split_face(hhh, next(next(hhh, tmesh), tmesh), tmesh);
+              FT l = edge_length(hf, tmesh, parameters::vertex_point_map(vpm).geom_traits(gt));
+              if (l >= threshold)
+              {
+                double nb_subsegments = std::ceil(l / threshold);
+                to_split_new.emplace_back(edge(hf,tmesh), nb_subsegments);
+              }
+            }
+          }
+        }
+      }
+      to_split.swap(to_split_new);
+    }
+    nb_vertices = vertices(tmesh).size();
+  }
+
+  while (nb_clusters > nb_vertices * vertex_count_ratio)
+  {
+    FT curr_factor = nb_clusters / (nb_vertices * vertex_count_ratio);
+    int subdivide_steps = (std::max)((int)ceil(log(curr_factor) / log(4)), 0);
```

Comment:
`log(FT)` ?

Reply:
a class model of `Kernel`, with `GT::FT` being either `float` or `double`.